File: Leetcode/1675 Minimize Deviation in Array.java
import java.util.Comparator;
import java.util.PriorityQueue;

class Solution {
    public int minimumDeviation(int[] nums) {
        var maxHeap = new PriorityQueue<Integer>(Comparator.reverseOrder());
        int min = Integer.MAX_VALUE;
        for (int i = 0; i < nums.length; i++) {
            nums[i] = nums[i] % 2 == 1 ? nums[i] * 2 : nums[i];
            min = Math.min(min, nums[i]);
            maxHeap.add(nums[i]);
        }
        int deviation = Integer.MAX_VALUE;
        while (maxHeap.peek() % 2 == 0) {
            deviation = Math.min(deviation, maxHeap.peek() - min);
            int updatedMax = maxHeap.poll() / 2;
            min = Math.min(min, updatedMax);
            maxHeap.add(updatedMax);
        }
        deviation = Math.min(deviation, maxHeap.peek() - min);
        return deviation;
    }
}

File: Leetcode/1675 Minimize Deviation in Array.py
import heapq


class Solution:
    def minimumDeviation(self, nums):
        min_value = float("inf")
        for i in range(len(nums)):
            if nums[i] % 2 == 1:
                nums[i] *= 2
            min_value = min(min_value, nums[i])

        max_heap = [-n for n in nums]
        heapq.heapify(max_heap)

        deviation = min_value - max_heap[0]
        while max_heap[0] % 2 == 0:
            max_val = -heapq.heappop(max_heap)
            deviation = min(deviation, max_val - min_value)
            reduced = max_val//2
            min_value = min(min_value, reduced)
            heapq.heappush(max_heap, -reduced)

        max_val = -heapq.heappop(max_heap)
        deviation = min(deviation, max_val - min_value)
        return deviation
